File: WordPress/WordPressTest/WPUserAgentTests.m
#import <XCTest/XCTest.h>

#import "WPUserAgent.h"
@import WebKit;

static NSString* const WPUserAgentKeyUserAgent = @"UserAgent";

@interface WPUserAgentTests : XCTestCase
@end

@implementation WPUserAgentTests

- (NSString *)currentUserAgentFromUserDefaults
{
    return [[NSUserDefaults standardUserDefaults] objectForKey:WPUserAgentKeyUserAgent];
}

- (NSString *)currentUserAgentFromWebView
{
    return [WKWebView userAgent];
}

- (void)testWordPressUserAgent
{
    NSString *appVersion = [[[NSBundle mainBundle] infoDictionary] objectForKey:@"CFBundleShortVersionString"];
    NSString *defaultUserAgent = [WPUserAgent defaultUserAgent];
    NSString *expectedUserAgent = [NSString stringWithFormat:@"%@ wp-iphone/%@", defaultUserAgent, appVersion];

    XCTAssertEqualObjects([WPUserAgent wordPressUserAgent], expectedUserAgent);
}

- (NSRegularExpression *)webkitUserAgentRegex
{
    NSError *error = nil;
    NSRegularExpression *regex = [[NSRegularExpression alloc] initWithPattern:@"^Mozilla/5\\.0 \\([a-zA-Z]+; CPU [\\sa-zA-Z]+ [_0-9]+ like Mac OS X\\) AppleWebKit/605\\.1\\.15 \\(KHTML, like Gecko\\) Mobile/15E148$" options:0 error:&error];
    XCTAssertNil(error);
    return regex;
}

- (void)testUserAgentFormat
{
    NSRegularExpression *regex = [self webkitUserAgentRegex];
    NSString *userAgent = [WPUserAgent webViewUserAgent];
    XCTAssertEqual([regex numberOfMatchesInString:userAgent options:0 range:NSMakeRange(0, userAgent.length)], 1);
}

// If this test fails, it may mean `WKWebView` uses a user agent with an unexpected format (see `webkitUserAgentRegex`)
// and we may need to adjust `UserAgent.webkitUserAgent`'s implementation to match `WKWebView`'s user agent.
- (void)testWKWebViewUserAgentFormat
{
    NSRegularExpression *regex = [self webkitUserAgentRegex];
    // Please note: WKWebView's user agent may be different on different test device types.
    NSString *userAgent = [self currentUserAgentFromWebView];
    XCTAssertEqual([regex numberOfMatchesInString:userAgent options:0 range:NSMakeRange(0, userAgent.length)], 1);
}

- (void)testUseWordPressUserAgentInWebViews
{
    NSString *defaultUA = [WPUserAgent defaultUserAgent];
    NSString *wordPressUA = [WPUserAgent wordPressUserAgent];

    [[NSUserDefaults standardUserDefaults] removeObjectForKey:WPUserAgentKeyUserAgent];
    [[NSUserDefaults standardUserDefaults] registerDefaults:@{WPUserAgentKeyUserAgent: defaultUA}];

    XCTAssertEqualObjects([self currentUserAgentFromUserDefaults], defaultUA);
    XCTAssertEqualObjects([self currentUserAgentFromWebView], defaultUA);

    if (@available(iOS 17, *)) {
        XCTSkip("In iOS 17, WKWebView no longer reads User Agent from UserDefaults. Skipping while working on an alternative setup.");
    }

    [WPUserAgent useWordPressUserAgentInWebViews];
    XCTAssertEqualObjects([self currentUserAgentFromUserDefaults], wordPressUA);
    XCTAssertEqualObjects([self currentUserAgentFromWebView], wordPressUA);
}

- (void)testThatOriginalRemovalOfWPUseKeyUserAgentDoesntWork {
    if (@available(iOS 17, *)) {
        XCTSkip("In iOS 17, WKWebView no longer reads User Agent from UserDefaults. Skipping while working on an alternative setup.");
    }

    // get the original user agent
    NSString *originalUserAgentInWebView = [self currentUserAgentFromWebView];
    NSLog(@"OriginalUserAgent (WebView): %@", originalUserAgentInWebView);
    
    // set a new one
    [[NSUserDefaults standardUserDefaults] registerDefaults:@{WPUserAgentKeyUserAgent: @"new user agent"}];
    
    NSString *changedUserAgentInWebView = [self currentUserAgentFromWebView];
    NSLog(@"changedUserAgent (WebView): %@", changedUserAgentInWebView);

    // try to remove it using old method
    [[NSUserDefaults standardUserDefaults] registerDefaults:@{}];
    [[NSUserDefaults standardUserDefaults] removeObjectForKey:WPUserAgentKeyUserAgent];
    
    NSString *shouldBeOriginalInWebView = [self currentUserAgentFromWebView];
    NSLog(@"shouldBeOriginal (WebView): %@", shouldBeOriginalInWebView);
    
    XCTAssertNotEqualObjects(originalUserAgentInWebView, shouldBeOriginalInWebView);
}

- (void)testThatCallingFromAnotherThreadWorks {
    // get the original user agent
    dispatch_sync(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
        XCTAssertNoThrow([WPUserAgent wordPressUserAgent], @"Being called from out of main thread should work");
    });
}

- (void)testThatRegistarDefaultJustAdds {
    NSDictionary * registrationDomain = [[NSUserDefaults standardUserDefaults] volatileDomainForName:NSRegistrationDomain];
    [[NSUserDefaults standardUserDefaults] registerDefaults:@{WPUserAgentKeyUserAgent: @(0)}];
    NSDictionary * changedRegistrationDomain = [[NSUserDefaults standardUserDefaults] volatileDomainForName:NSRegistrationDomain];

    XCTAssertTrue((registrationDomain.count == changedRegistrationDomain.count) || ((registrationDomain.count +1) == changedRegistrationDomain.count), "It should add or reset");
}


@end

File: WordPress/WordPressTest/WPUserAgentTests.swift
import Testing
import WebKit
@testable import WordPress

class WPWPUserAgentTests {

    @Test
    func wordPressUserAgentValue() throws {
        let userDefaults = UserDefaults.standard
        let appVersion = try #require(Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String)
        let defaultUserAgent = TemporaryWPUserAgent.defaultUserAgent(userDefaults: userDefaults)
        let expectedUserAgent = String.init(format: "%@ wp-iphone/%@", defaultUserAgent, appVersion)

        #expect(TemporaryWPUserAgent.wordPressUserAgent(userDefaults: userDefaults) == expectedUserAgent)
    }

    @Test
    func usesWordPressUserAgentInWebViews() throws {
        if #available(iOS 17, *) { // #available cannot go as an argument in @Test(.enabled(if: ..))
            print("In iOS 17, WKWebView no longer reads User Agent from UserDefaults. Skipping while working on an alternative setup.")
            return
        }

        let userDefaults = UserDefaults.standard
        let defaultUserAgent = TemporaryWPUserAgent.defaultUserAgent(userDefaults: userDefaults)
        let wordPressUserAgent = TemporaryWPUserAgent.wordPressUserAgent(userDefaults: userDefaults)

        // FIXME: Is this necessary?
        // See original implementation at
        // https://github.com/wordpress-mobile/WordPress-iOS/blob/a6eaa7aa8acb50828449df2d3fccaa50d7def821/WordPress/WordPressTest/WPUserAgentTests.m#L57-L75
        userDefaults.removeObject(forKey: TemporaryWPUserAgent.userAgentKey)
        userDefaults.register(defaults: [TemporaryWPUserAgent.userAgentKey: defaultUserAgent])

        WPUserAgent.useWordPressInWebViews()

        #expect(try currentUserAgent(userDefaults: userDefaults) == wordPressUserAgent)
        #expect(try currentUserAgentFromWebView() == wordPressUserAgent)
    }

    // FIXME: Is there even a point in testing for no throws when the method does not throw?
    // See original implementation at
    // https://github.com/wordpress-mobile/WordPress-iOS/blob/a6eaa7aa8acb50828449df2d3fccaa50d7def821/WordPress/WordPressTest/WPUserAgentTests.m#L102-L107
    @Test
    func accessingWordPressUserAgentOutsideMainThread() {
        #expect(throws: Never.self, "Accessing outside the main thread should work") {
            DispatchQueue.global(qos: .background).sync {
                TemporaryWPUserAgent.wordPressUserAgent(userDefaults: .standard)
            }
        }
    }

    func currentUserAgent(userDefaults: UserDefaults) throws -> String {
        try #require(userDefaults.object(forKey: TemporaryWPUserAgent.userAgentKey) as? String)
    }

    func currentUserAgentFromWebView() throws -> String {
        try #require(WKWebView.userAgent())
    }
}
